crates/providers/src/provider.rs

```diff
@@ -31,223 +33,432 @@ pub struct Provider<T: Transport = BoxTransport> {
     from: Option<Address>,
 }
 
+/// Temporary Provider trait to be used until the new Provider trait with
+/// the Network abstraction is stable.
+/// Once the new Provider trait is stable, this trait will be removed.
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+#[auto_impl(&, Arc, Box)]
```

Comment:
this can be auto impld on &mut and Rc

Reply:
done on https://github.com/alloy-rs/alloy/pull/20/commits/f767a997f24b21f396ce660e12507f4fa064b8bb